### PR TITLE
Exponential Backoff

### LINKED ISSUE #1499: Reconnect loop is far to quick
The 3 seconds reconnect DDoses the server is the server is down for a while and comes back up

### PR DESCRIPTION
**Pull request by 1-alex98**
Fixes #1499

### CODE REVIEW

File: src/main/java/com/faforever/client/game/GameService.java [hunk 1]
```diff
@@ -200,6 +204,7 @@ public GameService(ClientProperties clientProperties,
         item -> new Observable[]{item.statusProperty(), item.getTeams()}
     );
     JavaFxUtil.attachListToMap(games, uidToGameInfoBean);
+    JavaFxUtil.addListener(gameRunning, (observable, oldValue, newValue) -> reconnectService.setGameRunning(newValue));
```
Comment (Brutus5000) [inline src/main/java/com/faforever/client/game/GameService.java]: I'm a bit confused. We have two services talking to each other. No UI involeved. Why are we using JavaFxUtil here? We should keep out JavaFx stuff wherever possible, don't we?

File: src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java [hunk 2]
```diff
@@ -164,6 +165,8 @@
   private Socket fafServerSocket;
   private CompletableFuture<List<Avatar>> avatarsFuture;
   private CompletableFuture<List<IceServer>> iceServersFuture;
+  @Setter
```
Comment (Brutus5000) [inline src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java]: I would suggest to change this state via Events on the message bus.

File: src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java [hunk 3]
```diff
@@ -247,15 +250,17 @@ protected Void call() throws Exception {
 
             logger.info("FAF server connection established");
             Platform.runLater(() -> connectionState.set(ConnectionState.CONNECTED));
+            reconnectService.resetConnectionFailures();
```
Comment (Brutus5000) [inline src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java]: This also belongs in the reconnect service (and could be also be triggered by events on the message bus?)

File: src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java [hunk 4]
```diff
@@ -247,15 +250,17 @@ protected Void call() throws Exception {
...
             blockingReadServer(fafServerSocket);
           } catch (IOException e) {
             Platform.runLater(() -> connectionState.set(ConnectionState.DISCONNECTED));
             if (isCancelled()) {
               logger.debug("Connection to FAF server has been closed");
             } else {
-              logger.warn("Lost connection to FAF server, trying to reconnect in " + RECONNECT_DELAY / 1000 + "s", e);
-              Thread.sleep(RECONNECT_DELAY);
+              logger.warn("Lost connection to Server", e);
+              reconnectService.incrementConnectionFailures();
```
Comment (Brutus5000) [inline src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java]: The reconnect service should track that itself

File: src/main/java/com/faforever/client/remote/ReconnectService.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+
+import java.util.Random;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+@Service
+@Slf4j
+public class ReconnectService {
```
Comment (Brutus5000) [inline src/main/java/com/faforever/client/remote/ReconnectService.java]: I am unclear about responsibilities here. I would assume a reconnect service is responsible for invoking the reconnect. But as it seems it is merely a simple util.

File: src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java [hunk 6]
```diff
@@ -247,15 +250,17 @@ protected Void call() throws Exception {
...
 
             blockingReadServer(fafServerSocket);
           } catch (IOException e) {
             Platform.runLater(() -> connectionState.set(ConnectionState.DISCONNECTED));
             if (isCancelled()) {
               logger.debug("Connection to FAF server has been closed");
             } else {
-              logger.warn("Lost connection to FAF server, trying to reconnect in " + RECONNECT_DELAY / 1000 + "s", e);
-              Thread.sleep(RECONNECT_DELAY);
+              logger.warn("Lost connection to Server", e);
```
Comment (Brutus5000) [inline src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java]: Can we switch to `@Slf4J` in this class?

**Review by Brutus5000 — CHANGES_REQUESTED** [review]

File: src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java (hunk 2 above)
Comment (Geosearchef) [reply src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java]: An event here could be really useful for other cases later on aswell.

File: src/main/java/com/faforever/client/remote/ReconnectService.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    connectionFailures++;
+  }
+
+  @VisibleForTesting
+  long getReconnectTimeOut() {
+    if (gameRunning) {
+      return RECONNECT_DELAY_FACTOR;
+    }
+    double timeToWait = RECONNECT_DELAY_FACTOR * Math.pow(2, connectionFailures);
+    timeToWait = timeToWait > RECONNECT_DELAY_MAX ? RECONNECT_DELAY_MAX : timeToWait;
```
Comment (Geosearchef) [inline src/main/java/com/faforever/client/remote/ReconnectService.java]: Use Math.min()

File: src/main/java/com/faforever/client/remote/ReconnectService.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+   * The maximum time somebody will ever wait to reconnect. Though don't worry in the case you are in a game it will
+   * always be RECONNECT_DELAY_FACTOR .
+   */
+  @VisibleForTesting
+  static final long RECONNECT_DELAY_MAX = 80000;
+  /**
+   * The time we start with waiting till we reconnect. After failures the time will increase exponentially.
+   */
+  private static final long RECONNECT_DELAY_FACTOR = 3000;
+  private static final Random RECONNECT_RANDOM = new Random(System.currentTimeMillis());
```
Comment (Geosearchef) [inline src/main/java/com/faforever/client/remote/ReconnectService.java]: No need to initialize with system time. Also no constant, although final, pls do not write in caps.

File: src/main/java/com/faforever/client/remote/ReconnectService.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+  }
+
+  @VisibleForTesting
+  long getReconnectTimeOut() {
+    if (gameRunning) {
+      return RECONNECT_DELAY_FACTOR;
+    }
+    double timeToWait = RECONNECT_DELAY_FACTOR * Math.pow(2, connectionFailures);
+    timeToWait = timeToWait > RECONNECT_DELAY_MAX ? RECONNECT_DELAY_MAX : timeToWait;
+    return (long) (RECONNECT_RANDOM.nextFloat() * timeToWait);
```
Comment (Geosearchef) [inline src/main/java/com/faforever/client/remote/ReconnectService.java]: Consider just using Math.random(), it's static and has an internal LCPRNG.

**Review by Geosearchef — CHANGES_REQUESTED** [review]

File: src/main/java/com/faforever/client/remote/ReconnectService.java (hunk 9 above)
Comment (1-alex98) [reply src/main/java/com/faforever/client/remote/ReconnectService.java]: Why and of cause I know what this means 😛

File: src/main/java/com/faforever/client/remote/ReconnectService.java (hunk 8 above)
Comment (1-alex98) [reply src/main/java/com/faforever/client/remote/ReconnectService.java]: SURE this is a constant.... THAT_IS_HOW_IT_IS_DONE

Comment (1-alex98) [reply src/main/java/com/faforever/client/remote/ReconnectService.java]: U are right about init tho

File: src/main/java/com/faforever/client/game/GameService.java (hunk 1 above)
Comment (1-alex98) [reply src/main/java/com/faforever/client/game/GameService.java]: It is for thread safety. Won't harm and should always be used

File: src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java (hunk 3 above)
Comment (1-alex98) [reply src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java]: as u please

File: src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java (hunk 4 above)
Comment (1-alex98) [reply src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java]: I put it on the event bus then

File: src/main/java/com/faforever/client/remote/ReconnectService.java (hunk 5 above)
Comment (1-alex98) [reply src/main/java/com/faforever/client/remote/ReconnectService.java]: Renamed it

File: src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java (hunk 6 above)
Comment (1-alex98) [reply src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java]: ok

File: src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java (hunk 3 above)
Comment (1-alex98) [reply src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java]: actually no, events are bad here as they are async and create the need for synchronization

File: src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java (hunk 4 above)
Comment (1-alex98) [reply src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java]: actually no, events are bad here as they are async and create the need for synchronization

**Comment by 1-alex98** [comment]
@Brutus5000 if u want this in next release tell me if it is ok? Or what to change

**Comment by Brutus5000** [comment]
I don't get your point of problems with the async events as the whole connectivity topic is async (or at least should be? disconnects can happen anytime). But apart from that it looks fine.

**Comment by 1-alex98** [comment]
hard to explain but it is an issue here

**Pull request merged**